crowdnotifier-sdk/src/main/java/org/crowdnotifier/android/sdk/utils/CryptoUtils.java

```diff
@@ -68,68 +71,156 @@ public EncryptedVenueVisit getEncryptedVenueVisit(long arrivalTime, long departu
 
 		for (EncryptedVenueVisit venueVisit : venueVisits) {
 
-			byte[] computed_h = new byte[KEY_BYTES];
-			int result = sodium.crypto_scalarmult(computed_h, eventInfo.getSecretKey(), venueVisit.getEphemeralPublicKey());
-			if (result != 0) { throw new RuntimeException("crypto_scalarmult returned a value != 0"); }
+			G1 secretKeyForIdentity = new G1();
+			secretKeyForIdentity.deserialize(eventInfo.getSecretKeyForIdentity());
 
-			if (!Arrays.equals(computed_h, venueVisit.getTag())) {
-				continue;
-			}
+			byte[] msg_p = decryptInternal(
+					new EncryptedData(venueVisit.getC1(), venueVisit.getC2(), venueVisit.getC3(), venueVisit.getNonce()),
+					secretKeyForIdentity, eventInfo.getIdentity());
+			if (msg_p == null) continue;
 
-			byte[] gR = new byte[KEY_BYTES];
-			result = sodium.crypto_scalarmult_base(gR, eventInfo.getSecretKey());
-			if (result != 0) { throw new RuntimeException("crypto_scalarmult_base returned a value != 0"); }
+			Payload payload = new Gson().fromJson(new String(msg_p), Payload.class);
 
-			byte[] tConcatAux = new byte[venueVisit.getEncryptedPayload().length - SEAL_BYTES];
-			result = sodium.crypto_box_seal_open(tConcatAux,venueVisit.getEncryptedPayload(),
-					venueVisit.getEncryptedPayload().length,gR,eventInfo.getSecretKey());
-			if (result != 0) { throw new RuntimeException("crypto_box_seal_open returned a value != 0"); }
+			byte[] decryptedMessage = crypto_secretbox_open_easy(payload.getNotificationKey(), eventInfo.getEncryptedMessage(),
+					eventInfo.getNonce());
 
-			byte[] t = Arrays.copyOfRange(tConcatAux, 0, 32);
-			byte[] aux = Arrays.copyOfRange(tConcatAux, 32, tConcatAux.length);
+			String decryptedMessageString = new String(decryptedMessage);
 
-			byte[] tConcatR2 = concatenate(t, eventInfo.getR2());
-			byte[] skP = new byte[KEY_BYTES];
-			result = sodium.crypto_hash_sha256(skP, tConcatR2, tConcatR2.length);
-			if (result != 0) { throw new RuntimeException("crypto_hash_sha256 returned a value != 0"); }
+			ExposureEvent exposureEvent = new ExposureEvent(venueVisit.getId(), payload.getArrivalTime(),
+					payload.getDepartureTime(), decryptedMessageString);
 
+			exposureEvents.add(exposureEvent);
+		}
+		return exposureEvents;
+	}
 
-			byte[] venuePublicKey = new byte[Box.PUBLICKEYBYTES];
-			byte[] venuePrivateKey = new byte[Box.SECRETKEYBYTES];
-			result = sodium.crypto_box_seed_keypair(venuePublicKey, venuePrivateKey, skP);
-			if (result != 0) { throw new RuntimeException("crypto_box_seed_keypair returned a value != 0"); }
+	public byte[] decryptInternal(EncryptedData encryptedData, G1 secretKeyForIdentity, byte[] identity) {
+		G2 c1 = new G2();
+		c1.deserialize(encryptedData.getC1());
 
+		GT gt_temp = new GT();
+		Mcl.pairing(gt_temp, secretKeyForIdentity, c1);
 
-			if (!Arrays.equals(venuePrivateKey, eventInfo.getSecretKey())) {
-				continue;
-			}
+		byte[] hash = crypto_hash_sha256(gt_temp.serialize());
+		byte[] x_p = xor(encryptedData.getC2(), hash);
```

Comment:
bit of a naming discrepancy between `x_p` (which is x' ("x prime") in the paper) and `nonceX` for encrypt. Maybe just call the second one `x` and this one `x_p` for consistency